packages/middleware/src/apiClientFactory/index.ts

```diff
@@ -20,89 +20,91 @@ const apiClientFactory = <
   factoryParams: ApiClientFactoryParams<ALL_SETTINGS, ALL_FUNCTIONS>
 ): ApiClientFactory<any, ALL_FUNCTIONS> => {
   const createApiClient: CreateApiClientFn<any, ALL_FUNCTIONS> =
-    function createApiClient(
+    async function createApiClient(
       this: CallableContext<ALL_FUNCTIONS>,
       config,
       customApi: Record<string, any> = {}
     ) {
       const rawExtensions: ApiClientExtension<ALL_FUNCTIONS>[] =
         this?.middleware?.extensions || [];
 
-      const lifecycles = rawExtensions
-        .filter((extension): extension is ExtensionWith<"hooks"> =>
-          isFunction(extension?.hooks)
-        )
-        .map(({ hooks }) =>
-          hooks(this?.middleware?.req, this?.middleware?.res)
-        );
+      const lifecycles = await Promise.all(
+        rawExtensions
+          .filter((extension): extension is ExtensionWith<"hooks"> =>
+            isFunction(extension?.hooks)
+          )
+          .map(async ({ hooks }) =>
+            hooks(this?.middleware?.req, this?.middleware?.res)
+          )
+      );
 
-      const _config = lifecycles
+      const _config = await lifecycles
         .filter((extension): extension is ExtensionHookWith<"beforeCreate"> =>
           isFunction(extension?.beforeCreate)
         )
-        .reduce(
-          (configSoFar, extension) =>
-            extension.beforeCreate({ configuration: configSoFar }),
-          config
-        );
+        .reduce(async (configSoFar, extension) => {
+          const resolvedConfig = await configSoFar;
+          return extension.beforeCreate({ configuration: resolvedConfig });
```

Comment:
await on lifecycles seems fine but I think you missed async behaviour on hooks like `beforeCreate`, `afterCreate`, `beforeCall` and `afterCall`